### ISSUE TITLE
UX - allow a site editor to add descriptions to custom blocks.

### ISSUE THREAD

**Issue by Graham-72** [backdrop/backdrop-issues#2722]
As a site developer I want to make it easier for my site editors to understand the purpose of individual custom blocks, particularly when there are several in the site. At the moment the title is of some help but there is existing provision in the block module for a description field and I propose that, for custom blocks, this should be editable and should be displayed both on the custom blocks listing and on the manage blocks page in layouts.

At the moment once a custom block is made reusable, the description field contains the default wording 'A reusable custom block'. This proposal would make this an editable field containing custom content.

---

~PR by @Graham-72: https://github.com/backdrop/backdrop/pull/1908 includes amendments proposed by @klonos.~
~~PR by @Graham-72: https://github.com/backdrop/backdrop/pull/2123~~
PR by @jenlampton https://github.com/backdrop/backdrop/pull/2149

**Comment by klonos** [comment]
Just gave the PR sandbox a spin and I found a flaw with the UX workflow: While the description field becomes available for *existing* reusable custom blocks, it is not made available when creating one. So one needs to either edit or create a new custom block, make it reusable, save the block, and then re-edit the block in order to add a description.

Ideally, how this should work is that the description field becomes available when checking the "Make this block reusable" checkbox, but then that presents another issue: The description field makes sense below the block title field, but if the user has scrolled down to the "Make this block reusable" checkbox, there is a chance that they miss the field if that is shown via `#states` outside their viewport. Hmm. We can either move the description field below the admin label field and have it be show via `#states`, or have it be available below the block title field at all times (even for non-reusable blocks).

**Comment by Graham-72** [comment]
@klonos Thank you for testing this but I am a bit puzzled by your result.

Before applying the patch the description field is not available when creating a new custom block but after applying the patch it is. Here are  my screenshots of before and after:

BEFORE
![screenshot-www cgwebdesign co uk-2017-06-07-07-33-51](https://user-images.githubusercontent.com/10559274/26865742-f8851010-4b55-11e7-8f99-ae79d905675d.png)

AFTER
![screenshot-www cgwebdesign co uk-2017-06-07-07-46-25](https://user-images.githubusercontent.com/10559274/26865755-09591f9e-4b56-11e7-97e2-9c1a76cea63b.png)

I obtained these from the menu link admin/structure/block/add
Is there perhaps some other path for creating a custom block? Please tell me how you obtained your result.

**Comment by olafgrabienski** [comment]
> While the description field becomes available for existing reusable custom blocks, it is not made available when creating one.

I can confirm this restriction which happens when I create a custom block via the Layouts UI, e.g. adding a custom block to a region via `admin/structure/layouts/manage/default`.

When I create a custom block via `admin/structure/block/add`, I can see the description field and fill something in but the description is not saved. When I configure this block and add the description again, it is saved finally.

**Comment by Graham-72** [comment]
Thanks for the feedback on this. I will do some more work on it.

**Comment by klonos** [comment]
Yep, I have tried this by hitting the "Add block" button available in a region in the manage blocks UI, and then selected "Custom block".

It is quite confusing why we are using two different forms for the same thing. One has the "Admin label" field required and visible all the time, while the other has it optional and only shown via `#states` when the "Make this block reusable" checkbox is ticked. So adding a custom block via the manage blocks UI allows you to make the block reusable, while adding a custom block from `/admin/structure/block/add` does not give you that option. A matter for a separate ticket. Putting it here for now.

**Comment by Graham-72** [comment]
> So adding a custom block via the manage blocks UI allows you to make the block reusable, while adding a custom block from /admin/structure/block/add does not give you that option. A matter for a separate ticket. Putting it here for now.

@klonos I have done some more work on the PR to provide the description field in both cases. I think it is the case that when creating a custom block using /admin/structure/block/add it is always a reusable block and this seems sensible to me.

Also, the description is now saved correctly in config when a new custom block is created.

**Comment by klonos** [comment]
> I think it is the case that when creating a custom block using /admin/structure/block/add it is always a reusable block and this seems sensible to me.

Now that you said it, I find it sensible too. Another question (not specifically related to this issue and the PR): Is there any reason why the custom block "Body content" field is required in `admin/structure/block/add` while it is optional in the add/edit custom block dialog in the Block management UI?

**Comment by Graham-72** [comment]
Is it perhaps so that, when working from the block management UI, you can create a placeholder block with no content yet? 

I have never yet created a custom block this way, always using the more obvious (to me) /admin/structure/block/add.

**Comment by klonos** [comment]
Gave the PR sandbox a quick spin and it seems to be working as expected now. Good job @Graham-72! One minor nit: 

I also took a quick look at the PR code and left a few comments re the order of fields and some other minor things here and there. Other than that, RTBC from me 👍 (but someone else should also take a look at the code)

PS: I guess that this PR will clash with mine in #2731 (both change the same parts of `block.text.inc`), but I'll revise mine once this is merged.

PPS: I would love to see what you guys think about #2733

**Comment by Graham-72** [comment]
@klonos thanks for looking at this and for your comments on the PR code. Yes, I will change the field weights, and yes we do need the fix in your #2731.

**Comment by klonos** [comment]
@Graham-72 NP. Lets get your changes in first, since you filed this issue and PR before mine. Then I'll revise my PR.

I really love it when we fix tiny bits of the core UI here and there. It might seem as extremely pedantic and OCD-ish at times, but it's the attention to small UX details that makes Backdrop a better D7 😄 ...plus, it gives us beginners something to work on while learning and leaves our experienced devs free to deal with the more important tickets. Teamwork++

**Comment by klonos** [comment]
...talking about being pedantic, should we also fix the table headers in `admin/structure/block/list` while we're at it (pretty much what we do in `admin/structure/layouts/list`)?

- add a `priority-low` class in the description column (both `th`s and `td`s), so that it gets hidden on mobile.
- add a `custom-block-operations` class in the operations column and give it a `text-align: center` because:

![image](https://user-images.githubusercontent.com/2423362/27057397-5cca3d1c-500f-11e7-86f3-46174347cee6.png)

**Comment by quicksketch** [comment]
Looking great here, but this PR needs a bit more work, as comments in the PR state. We've got some PHP notices cropping up in the current code.

**Comment by Graham-72** [comment]
@quicksketch @klonos Thanks for your comments. I have now made one further small change to the PR and I don't know/understand what more needs to be done.

**Comment by opi** [comment]
Tested and approved https://github.com/backdrop/backdrop/pull/1908#pullrequestreview-109493549

**Note** PR need to be rebased before merge.

**Comment by olafgrabienski** [comment]
I've had a look at the block descriptions on the sandbox site, both on the *Custom blocks* page and on a *Layout* page. Looks good, great work!

As a last thing, on the *Custom blocks* page we could "add a `priority-low` class in the description column (both `th`s and `td`s), so that it gets hidden on mobile", as @klonos suggested. Makes sense, in my opinion.

**Comment by jenlampton** [comment]
I also love this change, but I'm not happy with the field label being `Block description`. I'd prefer `Admin description` so that we're consistent with the `Admin label`. 

![screen shot 2018-04-24 at 11 37 05 pm](https://user-images.githubusercontent.com/397895/39229651-70c1b050-4818-11e8-831a-c46c41edab37.png)

I'm going to take a stab at a PR.

**Comment by jenlampton** [comment]
I pushed a new PR that looks like it could use one more sandbox-review. I think it's close!

**Comment by olafgrabienski** [comment]
Wanted to have a look at it, but I don't see a sandbox website on https://github.com/backdrop/backdrop/pull/2149.

**Comment by opi** [comment]
@jenlampton you missed one label when changing from "Block description" to "Admin description" (see https://github.com/backdrop/backdrop/pull/2149/files#r184474162)

Consider this as an approval anyway, reviewed & tested, it works as expected. Thanks everyone

**Comment by quicksketch** [comment]
I've merged in https://github.com/backdrop/backdrop/pull/2149 into 1.x for 1.10. I did not merge it into 1.9.x (though I could as this is a UX improvement). With 1.10 around the corner I hope that is acceptable.
